actors/builtin/market/market_test.go

```diff
@@ -1434,11 +1440,80 @@ func TestCronTick(t *testing.T) {
+	owner := tutil.NewIDAddr(t, 101)
+	provider := tutil.NewIDAddr(t, 102)
+	worker := tutil.NewIDAddr(t, 103)
+	client := tutil.NewIDAddr(t, 104)
+	mAddrs := &minerAddrs{owner, worker, provider}
+
+	startEpoch := abi.ChainEpoch(50)
+	endEpoch := startEpoch + 200*builtin.EpochsInDay
+	sectorExpiry := endEpoch + 1
+
+	t.Run("a random epoch in chosen as the next cron tick for a deal only at the startEpoch", func(t *testing.T) {
+		rt, actor := basicMarketSetup(t, owner, provider, worker, client)
+		dealId := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch, endEpoch)
+		d := actor.getDealProposal(rt, dealId)
+
+		rt.SetEpoch(startEpoch - 1)
+		actor.activateDeals(rt, sectorExpiry, provider, d.StartEpoch-1, dealId)
+
+		// first cron tick will schedule the deal for the given random epoch
+		rt.SetEpoch(startEpoch)
+		actor.cronTick(rt, d.StartEpoch, cronFirstTimeDeal{d, startEpoch + 2})
+
+		// a cron tick at 51 will not do anything
+		rt.SetEpoch(startEpoch + 1)
+		actor.cronTickNoChange(rt, client, provider, startEpoch+1)
+
+		// but a cron tick at 52 will make a payment and schedule the deal to be processed at the next deal Interval
+		current := startEpoch + 2
+		rt.SetEpoch(current)
+		pay, _ := actor.cronTickAndAssertBalances(rt, client, provider, startEpoch+2, dealId, undefinedEpoch)
+		require.EqualValues(t, big.Mul(big.NewInt(2), d.StoragePricePerEpoch), pay)
+
+		current = current + market.DealUpdatesInterval
+		rt.SetEpoch(current)
+		pay, _ = actor.cronTickAndAssertBalances(rt, client, provider, current, dealId, undefinedEpoch)
+		require.EqualValues(t, big.Mul(big.NewInt(market.DealUpdatesInterval), d.StoragePricePerEpoch), pay)
+	})
+
+	t.Run("deals are scheduled for expiry later than the end epoch", func(t *testing.T) {
+		rt, actor := basicMarketSetup(t, owner, provider, worker, client)
+		dealId := actor.generateAndPublishDeal(rt, client, mAddrs, startEpoch, endEpoch)
+		d := actor.getDealProposal(rt, dealId)
+
+		rt.SetEpoch(startEpoch - 1)
+		actor.activateDeals(rt, sectorExpiry, provider, d.StartEpoch-1, dealId)
+
+		// a cron tick at end epoch -1 schedules the deal for later than end epoch
+		curr := endEpoch - 1
+		rt.SetEpoch(curr)
+
+		duration := big.NewInt(int64(curr - startEpoch))
+		nextEpoch := endEpoch + 100
+		pay, _ := actor.cronTickAndAssertBalances(rt, client, provider, curr, dealId, nextEpoch)
+		require.EqualValues(t, big.Mul(duration, d.StoragePricePerEpoch), pay)
+
+		// cron tick at end epoch does NOT expire the deal
+		rt.SetEpoch(endEpoch)
+		actor.cronTickNoChange(rt, client, provider, endEpoch)
+		require.NotNil(t, actor.getDealProposal(rt, dealId))
+
+		// cron tick at nextEpoch expires the deal -> payment is ONLY for one epoch
+		rt.SetEpoch(nextEpoch)
+		pay, _ = actor.cronTickAndAssertBalances(rt, client, provider, nextEpoch, dealId, undefinedEpoch)
+		require.EqualValues(t, d.StoragePricePerEpoch, pay)
+		actor.assertDealDeleted(rt, dealId, d)
+	})
+}
+
```

Comment:
Since deal states are no longer definitely processed on their start or end epoch, we need higher level tests for new boundary conditions:
- activation after deal start epoch, but before it was processed (should fail)
- cron processing of deal after missed activation ↑
- termination of deal after it already expired, but before expiration was processed (should not slash)

Reply:
All three have been written. Please take a look at `TestRandomCronEpochDuringPublish`.